ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/engine/EngineNewPayloadV3.java

```diff
@@ -0,0 +1,130 @@
+ * specific language governing permissions and limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package org.hyperledger.besu.ethereum.api.jsonrpc.internal.methods.engine;
+
+import static org.hyperledger.besu.ethereum.api.jsonrpc.internal.methods.ExecutionEngineJsonRpcMethod.EngineStatus.INVALID;
+import static org.hyperledger.besu.ethereum.api.jsonrpc.internal.response.JsonRpcError.INVALID_PARAMS;
+import static org.hyperledger.besu.ethereum.api.jsonrpc.internal.response.JsonRpcError.UNSUPPORTED_FORK;
+
+import org.hyperledger.besu.consensus.merge.blockcreation.MergeMiningCoordinator;
+import org.hyperledger.besu.datatypes.VersionedHash;
+import org.hyperledger.besu.ethereum.ProtocolContext;
+import org.hyperledger.besu.ethereum.api.jsonrpc.RpcMethod;
+import org.hyperledger.besu.ethereum.api.jsonrpc.internal.parameters.EnginePayloadParameter;
+import org.hyperledger.besu.ethereum.api.jsonrpc.internal.response.JsonRpcErrorResponse;
+import org.hyperledger.besu.ethereum.api.jsonrpc.internal.response.JsonRpcResponse;
+import org.hyperledger.besu.ethereum.core.Transaction;
+import org.hyperledger.besu.ethereum.eth.manager.EthPeers;
+import org.hyperledger.besu.ethereum.mainnet.ProtocolSchedule;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+import io.vertx.core.Vertx;
+import org.apache.tuweni.bytes.Bytes32;
+
+public class EngineNewPayloadV3 extends AbstractEngineNewPayload {
+
+  private final ProtocolSchedule timestampSchedule;
+
+  public EngineNewPayloadV3(
+      final Vertx vertx,
+      final ProtocolSchedule timestampSchedule,
+      final ProtocolContext protocolContext,
+      final MergeMiningCoordinator mergeCoordinator,
+      final EthPeers ethPeers,
+      final EngineCallListener engineCallListener) {
+    super(
+        vertx, timestampSchedule, protocolContext, mergeCoordinator, ethPeers, engineCallListener);
+    this.timestampSchedule = timestampSchedule;
+  }
+
+  @Override
+  public String getName() {
+    return RpcMethod.ENGINE_NEW_PAYLOAD_V3.getMethodName();
+  }
+
+  @Override
+  protected Optional<JsonRpcResponse> validateForkSupported(
+      final Object reqId, final EnginePayloadParameter payloadParameter) {
+    var cancun = timestampSchedule.hardforkFor(s -> s.fork().name().equalsIgnoreCase("Cancun"));
+
+    if (cancun.isPresent() && payloadParameter.getTimestamp() >= cancun.get().milestone()) {
+      if (payloadParameter.getDataGasUsed() == null
+          || payloadParameter.getExcessDataGas() == null) {
+        return Optional.of(new JsonRpcErrorResponse(reqId, INVALID_PARAMS));
+      }
+    } else {
+      if (payloadParameter.getDataGasUsed() != null
+          || payloadParameter.getExcessDataGas() != null) {
+        return Optional.of(new JsonRpcErrorResponse(reqId, UNSUPPORTED_FORK));
+      }
+    }
+    return Optional.empty();
```

Comment:
should we check versioned hashes param here too?

Reply:
This is handled optionally in the parent class.